ProjFS.Mac/PrjFSKextTests/VirtualizationRootsTests.mm

```diff
@@ -264,4 +271,119 @@ - (void)testRegisterProviderForPath_ExistingRecycledRoot
+    shared_ptr<vnode> testFileVnode = self->testMountPoint->CreateVnodeTree(filePath);
+
+    PrjFSVirtualizationRootXAttrData rootXattr = {};
+    vector<uint8_t> rootXattrData(sizeof(rootXattr), 0x00);
+    memcpy(rootXattrData.data(), &rootXattr, rootXattrData.size());
+    repoRootVnode->xattrs.insert(make_pair(PrjFSVirtualizationRootXAttrName, rootXattrData));
+
+    VirtualizationRootHandle foundRoot = VirtualizationRoot_FindForVnode(&self->dummyTracer, PrjFSPerfCounter_VnodeOp_FindRoot, PrjFSPerfCounter_VnodeOp_FindRoot_Iteration, testFileVnode.get(), self->dummyVFSContext);
+
+    XCTAssertTrue(VirtualizationRoot_IsValidRootHandle(foundRoot));
+    
+    if (VirtualizationRoot_IsValidRootHandle(foundRoot))
+    {
+        XCTAssertEqual(s_virtualizationRoots[foundRoot].rootInode, repoRootVnode->GetInode());
+        
+        repoRootVnode->StartRecycling();
+        
+        shared_ptr<vnode> newRepoRootVnode = self->testMountPoint->CreateVnode(repoPath, VnodeCreationProperties { .inode = repoRootVnode->GetInode(), .type = VDIR });
+        
+        shared_ptr<vnode> otherTestFileVnode = self->testMountPoint->CreateVnodeTree(otherFilePath);
+        VirtualizationRootHandle foundRootForOther = VirtualizationRoot_FindForVnode(&self->dummyTracer, PrjFSPerfCounter_VnodeOp_FindRoot, PrjFSPerfCounter_VnodeOp_FindRoot_Iteration, otherTestFileVnode.get(), self->dummyVFSContext);
+        
+        XCTAssertEqual(foundRootForOther, foundRoot, "Despite recycling the repo root vnode, we should end up with the same handle");
+        XCTAssertEqual(s_virtualizationRoots[foundRootForOther].rootVNode, newRepoRootVnode.get(), "Vnode should be refreshed");
+        XCTAssertEqual(s_virtualizationRoots[foundRootForOther].rootVNodeVid, newRepoRootVnode->GetVid(), "Vnode VID should be refreshed");
+    }
+}
+
+- (void)testFindForVnode_DirectoryInUndetectedRoot
+{
+    const char* repoPath = "/Users/test/code/Repo";
+    const char* subdirPath = "/Users/test/code/Repo/some/nested/subdir";
+    
+    shared_ptr<vnode> repoRootVnode = self->testMountPoint->CreateVnodeTree(repoPath, VDIR);
+    shared_ptr<vnode> testSubdirVnode = self->testMountPoint->CreateVnodeTree(subdirPath, VDIR);
+
+    XCTAssertTrue(vnode_isdir(testSubdirVnode.get()));
+
+    PrjFSVirtualizationRootXAttrData rootXattr = {};
+    vector<uint8_t> rootXattrData(sizeof(rootXattr), 0x00);
+    memcpy(rootXattrData.data(), &rootXattr, rootXattrData.size());
+    repoRootVnode->xattrs.insert(make_pair(PrjFSVirtualizationRootXAttrName, rootXattrData));
+
+    VirtualizationRootHandle foundRoot = VirtualizationRoot_FindForVnode(&self->dummyTracer, PrjFSPerfCounter_VnodeOp_FindRoot, PrjFSPerfCounter_VnodeOp_FindRoot_Iteration, testSubdirVnode.get(), self->dummyVFSContext);
+    
+    XCTAssertTrue(VirtualizationRoot_IsValidRootHandle(foundRoot));
+    if (VirtualizationRoot_IsValidRootHandle(foundRoot))
```

Comment:
We've already asserted this is true, so no 'if' necessary

Reply:
Unfortunately, it is. If `XCTAssert*`s fail, they don't immediately terminate the current test, but keep going. If `foundRoot` here is not a valid handle (e.g. `RootHandle_None`, i.e. -1) we must not use it to look up entries in the array of roots because we'll go out of bounds. Hence the `if` block.

Reply:
Ah interesting!  Generally tests stop once they hit an assert.  I wonder if that a purposeful change in `XCTAssert `or if there is another method for it.  Either way in that case the 'if' makes sense.

Reply:
Yeah, I don't know why they designed it like that. I'm not aware of a non-continuing assertion in XCTest.framework, and didn't immediately spot any in `XCTAssertions.h`.

Reply:
Not something for this PR, but do you think it would be worth defining our own assert for unit tests that does return immediately if the assert fails?

Reply:
I've filed #936 for this and a bunch of other, similar `assert()` related problems.